src/ApplicationIO.cpp

```diff
@@ -88,23 +86,26 @@ namespace geopm
     void ApplicationIOImp::connect(void)
     {
         if (!m_is_connected) {
-            m_sampler->initialize();
-            m_rank_per_node = m_sampler->rank_per_node();
-            m_prof_sample.resize(m_sampler->capacity());
-            std::vector<int> cpu_rank = m_sampler->cpu_rank();
-            if (m_profile_io_sample == nullptr) {
-                m_epoch_regulator = std::make_shared<EpochRuntimeRegulatorImp>(m_rank_per_node, m_platform_io, m_platform_topo);
-                m_profile_io_sample = std::make_shared<ProfileIOSampleImp>(cpu_rank, m_epoch_regulator);
-                platform_io().get_profileio()->connect(m_profile_io_sample, m_epoch_regulator);
+            m_application_sampler.get_sampler()->initialize();
+            m_prof_sample.resize(m_application_sampler.get_sampler()->capacity());
+
+            if (m_application_sampler.get_io_sample() == nullptr) {
+                auto rank_per_node = m_application_sampler.get_sampler()->rank_per_node();
+                auto epoch_regulator = std::make_shared<EpochRuntimeRegulatorImp>(rank_per_node, m_platform_io, m_platform_topo);
+                m_application_sampler.set_regulator(epoch_regulator);
+
+                auto profile_io_sample = std::make_shared<ProfileIOSampleImp>(m_application_sampler); // This could probably be moved
```

Comment:
should just use singleton.  not sure if two objects having shared_ptr to each other will be a problem.